File: 2020/day02/main.cpp
#include <cassert>
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <cstdio>

namespace {

struct PasswordData {
    explicit PasswordData(int min, int max, char chr, std::string password)
        : min(min), max(max), chr(chr), password(std::move(password)) {
    }

    bool Valid01() const {
        int count = 0;
        for (auto c : password) {
            if (c == chr) {
                ++count;
            }
        }

        return count >= min && count <= max;
    }

    bool Valid02() const {
        return (password[min - 1] == chr && password[max - 1] != chr) || (password[min - 1] != chr && password[max - 1] == chr);
    }

    bool operator==(const PasswordData &other) const {
        return min == other.min && max == other.max && chr == other.chr && password == other.password;
    }

    int min;
    int max;
    char chr;
    std::string password;
};

std::vector<PasswordData> ParseInput(const std::vector<std::string> &inputs) {
    std::vector<PasswordData> ret;
    char password[256];
    for (const auto &input : inputs) {
        int min, max;
        char ch;
        sscanf(input.c_str(), "%d-%d %c: %s", &min, &max, &ch, password);

        ret.emplace_back(min, max, ch, password);
    }

    return ret;
}

int Solve01(const std::vector<PasswordData> &passwords) {
    int ret = 0;
    for (const auto &password : passwords) {
        if (password.Valid01()) {
            ++ret;
        }
    }

    return ret;
}

int Solve02(const std::vector<PasswordData> &passwords) {
    int ret = 0;
    for (const auto &password : passwords) {
        if (password.Valid02()) {
            ++ret;
        }
    }

    return ret;
}

void Test01() {
    // clang-format off
    std::vector<std::string> inputs {
        "1-3 a: abcde",
        "1-3 b: cdefg",
        "2-9 c: ccccccccc",
    };
    // clang-format on

    auto passwords = ParseInput(inputs);
    assert(passwords.size() == 3);
    assert((passwords[0] == PasswordData(1, 3, 'a', "abcde")));
    assert((passwords[1] == PasswordData(1, 3, 'b', "cdefg")));
    assert((passwords[2] == PasswordData(2, 9, 'c', "ccccccccc")));

    assert(passwords[0].Valid01());
    assert(!passwords[1].Valid01());
    assert(passwords[2].Valid01());

    assert(Solve01(passwords) == 2);
}

void Test02() {
    // clang-format off
    std::vector<std::string> inputs {
        "1-3 a: abcde",
        "1-3 b: cdefg",
        "2-9 c: ccccccccc",
    };
    // clang-format on

    auto passwords = ParseInput(inputs);
    assert(passwords[0].Valid02());
    assert(!passwords[1].Valid02());
    assert(!passwords[2].Valid02());

    assert(Solve02(passwords) == 1);
}

} // namespace

int main() {
    Test01();
    Test02();

    std::vector<std::string> inputs;
    std::string line;
    while (std::getline(std::cin, line)) {
        inputs.push_back(line);
    }

    auto passwords = ParseInput(inputs);

    auto answer01 = Solve01(passwords);
    assert(answer01 == 580);
    std::cout << "Part01: " << answer01 << std::endl;

    auto answer02 = Solve02(passwords);
    assert(answer02 == 611);
    std::cout << "Part02: " << answer02 << std::endl;

    return 0;
}

File: 2020/day02/rust/src/main.rs
use std::io::{self, Read};

#[derive(Debug)]
struct Line<'a> {
    min: usize,
    max: usize,
    chr: u8,
    message: &'a str,
}

impl<'a> Line<'a> {
    fn valid01(&self) -> bool {
        let c = char::from(self.chr);
        let count = self.message.matches(c).count();
        count >= self.min && count <= self.max
    }

    fn valid02(&self) -> bool {
        let b1 = self.message.as_bytes()[self.min - 1] == self.chr;
        let b2 = self.message.as_bytes()[self.max - 1] == self.chr;
        b1 && !b2 || !b1 && b2
    }
}

fn solve01(lines: &Vec<Line>) -> usize {
    lines.iter().filter(|&line| line.valid01()).count()
}

fn solve02(lines: &Vec<Line>) -> usize {
    lines.iter().filter(|&line| line.valid02()).count()
}

fn parse_input(input: &str) -> Vec<Line> {
    let mut ret: Vec<Line> = Vec::new();
    for line in input.lines() {
        let parts: Vec<&str> = line.split_whitespace().collect();
        assert!(parts.len() == 3);

        let counts: Vec<usize> = parts[0].split('-').flat_map(|s| s.parse().ok()).collect();
        ret.push(Line {
            min: counts[0],
            max: counts[1],
            chr: parts[1].as_bytes()[0],
            message: parts[2],
        });
    }

    ret
}

fn main() -> io::Result<()> {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input)?;

    let lines = parse_input(&input);
    let answer1 = solve01(&lines);
    assert!(answer1 == 580);

    let answer2 = solve02(&lines);
    assert!(answer2 == 611);

    println!("Part01: {}", answer1);
    println!("Part02: {}", answer2);
    Ok(())
}

#[test]
fn test01() {
    let input = r#"1-3 a: abcde
1-3 b: cdefg
2-9 c: ccccccccc"#;

    let lines = parse_input(input);
    assert_eq!(solve01(&lines), 2);
}
